script.service.kodi.callbacks/default.py

```diff
@@ -0,0 +1,316 @@
+    paths.append(os.path.join(paths[0], 'lib'))
+    paths.append(os.path.join(paths[1], 'usertasks'))
+    for path in paths:
+        if not os.path.isdir(path):
+            try:
+                os.mkdir(path)
+            except OSError:
+                pass
+    for path in paths[1:]:
+        fn = os.path.join(path, '__init__.py')
+        if not os.path.isfile(fn):
+            try:
+                with open(fn, mode='w') as f:
+                    f.writelines('')
+            except (OSError, IOError):
+                pass
+
+
+class MainMonitor(xbmc.Monitor):
+    def __init__(self, dispatcher, publishers):
+        super(MainMonitor, self).__init__()
+        self.dispatcher = dispatcher
+        self.publishers = publishers
+
+    def onSettingsChanged(self):
+        log(msg=_('Settings change detected - attempting to restart'))
+        for p in self.publishers:
+            p.abort(0.525)
+        self.dispatcher.abort(0.25)
+        start()
+
+
+def start():
+    global log
+    settings = Settings()
+    settings.getSettings()
+    kl = KodiLogger()
+    if settings.general['elevate_loglevel'] is True:
+        kl.setLogLevel(xbmc.LOGNOTICE)
+    else:
+        kl.setLogLevel(xbmc.LOGDEBUG)
+    log = kl.log
+    log(msg=_('Settings read'))
+    dispatcher = PubSub_Threaded.Dispatcher(interval=settings.general['TaskFreq'], sleepfxn=xbmc.sleep)
+    log(msg=_('Dispatcher initialized'))
+
+    subscriberfactory = SubscriberFactory(settings, kl)
+    subscribers = subscriberfactory.createSubscribers()
+    for subscriber in subscribers:
+        dispatcher.addSubscriber(subscriber)
+    publisherfactory = PublisherFactory(settings, subscriberfactory.topics, dispatcher, kl, debug)
+    publisherfactory.createPublishers()
+    publishers = publisherfactory.ipublishers
+
+    dispatcher.start()
+    log(msg=_('Dispatcher started'))
+
+    for p in publishers:
+        try:
+            p.start()
+        except threading.ThreadError:
+            raise
+    log(msg=_('Publisher(s) started'))
+    return dispatcher, publishers
+
+
+def main():
+    xbmc.log(msg=_('$$$ [kodi.callbacks] - Staring kodi.callbacks ver: %s') % str(__version__), level=xbmc.LOGNOTICE)
+    if branch != 'master':
+        xbmcaddon.Addon().setSetting('installed branch', branch)
+    dispatcher, publishers = start()
+    dispatcher.q_message(PubSub_Threaded.Message(PubSub_Threaded.Topic('onStartup')))
+    monitor = MainMonitor(dispatcher, publishers)
+    log(msg=_('Entering wait loop'))
+    monitor.waitForAbort()
+
+    # Shutdown tasks
+    dispatcher.q_message(PubSub_Threaded.Message(PubSub_Threaded.Topic('onShutdown'), pid=os.getpid()))
+    xbmc.sleep(500)
+    log(msg=_('Shutdown started'))
+    for p in publishers:
+        try:
+            p.abort()
+        except threading.ThreadError as e:
+            log(msg=_('Error aborting: %s - Error: %s') % (str(p), str(e)))
+    dispatcher.abort()
+    xbmc.sleep(1000)
```

Comment:
What I'm trying to point out here is that if you ever need to use sleep to "fix" a problem, you have a fundamental underlying issue will eventually cause shutdown problems. It doesn't matter how much anecdotal evidence you get that it 'works'. By using sleep you actually only making the problem worse for yourself because this time that's potentially wasted _only_ waiting for time to pass. You don't have to use Monitor or waitForAbort to solve this. You can use any event mechanism you want.

This is not just about this one sleep call, but all of them. I don't know any details of this script to be able to give any more specifics, but I assure you there is no problem that sleep "solves" that cannot be solved correctly with events. It's not a blocker, so I think any further discussion on the topic is best taken in the forums. 

Ps: these 5 seconds scripts gets to shut down is _shared_ cpu time. If you are unlucky you will get a few operations executed and you are out of time; which is why it's so important to shut down _fast_, not just within 5 seconds of wall clock time.